### DISCUSSION TITLE
Saving results in empty file

### DISCUSSION THREAD

**Discussion by greaka** [helix-editor/helix#8532] [Troubleshooting] [answered]
### Summary

Whenever I want to edit .rs files, I end up with an empty file.

Upon saving the buffer, I get the message "file modified by an external process. Use w! to overwrite". w! doesn't change the outcome.

It is 100% reproducible and only occurs on .rs files. Toml files among others are fine. I don't know if LSP can be a source of problems, but I can tell that this happens with LSP disabled/enabled/installed/uninstalled all the same.

I have this issue for almost a year now. I hoped for a magical fix, but now I'm here.
I only have this issue on one of my machines and am happy to help debug this.

### Reproduction Steps

 

I tried this:

1. `hx src/lib.rs`
2. `:w` or `:w!`

I expected this to happen:
The file is saved as I see it displayed.

Instead, this happened:
The content of both the file and the buffer are deleted.

### Helix log

<details><summary>~/.cache/helix/helix.log</summary>

```
2023-10-13T12:05:35.017 helix_view::editor [ERROR] editor error: file modified by an external process, use :w! to overwrite
```

Yes, this is the entire log.

</details>

### Platform

Arch Linux

### Terminal Emulator

wezterm 20230712-072601-f4abf8fd

### Helix Version

helix 23.05

**Comment by kirawi** [comment]
We use mtime to detect if the file was changed after opening the file. I suspect that it's something related to your system but I'm not familiar enough with linux to say.

**Comment by the-mikedavis** [comment]
Do you have any custom language configuration - anything in `~/.config/helix/languages.toml`?

**Reply by greaka** [reply]
Yes, I do have a custom language configuration. However, that one is present on all machines.
```
greaka@greaka-arch-desktop ~ % diff ~/.config/helix/ /git/dotfiles/helix
Common subdirectories: /home/greaka/.config/helix/themes and /git/dotfiles/helix/themes
greaka@greaka-arch-desktop ~ % /usr/bin/cat ~/.config/helix/languages.toml
[[language]]
name = "rust"
formatter = { command = "cargo", args = ["+nightly", "fmt"] }

[language-server.rust-analyzer.config.check]
command = "clippy"
greaka@greaka-arch-desktop ~ %
```

**Reply by the-mikedavis** [reply, marked as answer]
External formatters pass in the file through stdin and expect the formatted result through stdout. cargo-fmt currently doesn't support that: https://github.com/rust-lang/rustfmt/pull/5071.

**Reply by greaka** [reply]
Is there a workaround? My `rustfmt.toml` causes errors in helix without nightly

**Reply by greaka** [reply]
I removed the setting and even tried `hx -c /dev/null` but this doesn't fix the issue though.

**Reply by the-mikedavis** [reply]
You should format in a separate terminal/window with `cargo +nightly format`.

`-c` only affects the editor config (`config.toml`) and not the language configuration (`languages.toml`). You should remove that `formatter` line from the `languages.toml`. If it's still happening make sure you don't have a `.helix/languages.toml` that has the same configuration

**Reply by greaka** [reply]
I have a non-default xdg home directory that helix respects. Found and fixed.
